src/zenml/integrations/gcp/image_builders/gcp_image_builder.py

```diff
@@ -0,0 +1,229 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Google Cloud Builder image builder implementation."""
+
+from typing import TYPE_CHECKING, Any, Dict, Optional, Tuple, cast
+from urllib.parse import urlparse
+
+from google.cloud.devtools import cloudbuild_v1
+
+from zenml.client import Client
+from zenml.enums import StackComponentType
+from zenml.image_builders import BaseImageBuilder
+from zenml.integrations.gcp import GCP_ARTIFACT_STORE_FLAVOR
+from zenml.integrations.gcp.flavors import GCPImageBuilderConfig
+from zenml.integrations.gcp.google_credentials_mixin import (
+    GoogleCredentialsMixin,
+)
+from zenml.logger import get_logger
+from zenml.stack import StackValidator
+from zenml.utils.image_builder_context_uploader import (
+    ImageBuilderContextUploader,
+)
+
+if TYPE_CHECKING:
+    from zenml.container_registries import BaseContainerRegistry
+    from zenml.image_builders import BuildContext
+    from zenml.stack import Stack
+
+logger = get_logger(__name__)
+
+
+class GCPImageBuilder(BaseImageBuilder, GoogleCredentialsMixin):
+    """Google Cloud Builder image builder implementation."""
+
+    @property
+    def config(self) -> GCPImageBuilderConfig:
+        """The stack component configuration.
+
+        Returns:
+            The configuration.
+        """
+        return cast(GCPImageBuilderConfig, self._config)
+
+    @property
+    def validator(self) -> Optional["StackValidator"]:
+        """Validates the stack for the GCP Image Builder.
+
+        The GCP Image Builder requires a remote container registry to push the
+        image to, and a GCP Artifact Store to upload the build context, so
+        Cloud Build can access it.
+
+        Returns:
+            Stack validator.
+        """
+
+        def _validate_remote_components(stack: "Stack") -> Tuple[bool, str]:
+            assert stack.container_registry
+
+            if stack.container_registry.config.is_local:
```

Comment:
I'm wondering if we need additional checks to make sure the container registry is also in GCP? Or is there some way to let cloud build know about the registry credentials to make this work if the registry is outside the GCP project?

Reply:
My idea was that you could provide the authentication credentials when registering a container registry. Then, using these credentials and knowing the type of container registry, we could have some conditions to add an additional step to login to the container registry, so the image could be pushed from Google Cloud Build to external and private registries such as Github, ECR, etc.

Having that said, I think it's unlikely that anyone would use a stack in which the container registry is non-GCP and the image builder is GCP.

Reply:
Definitely agree, we'll be working on authentication credentials soon anyway after which we can hopefully lift this restriction and provide cloud build with access to any container registry registered in ZenML